### PR TITLE
[WIP] Added context for the command

### PR DESCRIPTION
**Pull request by dgonzalez**
Command context and command customizations to follow up with https://github.com/kubernetes/kubernetes/pull/33435.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34418)

### CODE REVIEW

**Comment by dgonzalez** [comment]
@fabianofranz @deads2k please, feedback on the context + CmdCustomizations. Is this what you were talking about?

This will leave the creation on commands to something like this:
```
func NewCmdCreate(f *cmdutil.Factory, out io.Writer, context * CmdContext) *cobra.Command {
   // The code for the command
}
```

on this way, we can factor all the urls into a map on the cmd.go (or similar) and be able to check for  rot URLs.

Please give me feedback.

**Comment by fabianofranz** [comment]
@dgonzalez yep, that's the idea. Note that once you already have `out` in `CmdContext` you don't need to pass it in the constructor, making it just

```
func NewCmdCreate(f *cmdutil.Factory, ctx *CmdContext) *cobra.Command {
  ...
}
```

File: pkg/kubectl/cmd/context.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+limitations under the License.
+*/
+package cmd
+
+import (
+  "io"
+)
+
+type CmdCustomizations struct {
+  baseCmd       string
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/context.go]: Yeah, all fields exposed.

File: pkg/kubectl/cmd/context.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  cmdName       string
+  cmdDetailsURL string
+}
+
+type CmdContext struct {
+  out            io.Writer
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/context.go]: Store all the three streams that commands can use: `In io.Reader`, `Out io.Writer`, and `ErrOut io.Writer`.

File: pkg/kubectl/cmd/context.go (hunk 1 above)
Comment (fabianofranz) [inline pkg/kubectl/cmd/context.go]: Other than `baseCmd` and `cmdName` you may also want to have `fullName`. Sometimes you have commands that live in the third or below level of the tree, for example `kubectl create quota`, in which case just the base and cmd name are not enough.

**Comment by fabianofranz** [comment]
Also, note that it would be too verbose to create the entire `CmdContext` struct every time I call the constructor of a command. Given most attributes repeat, we could have a `New` method that just takes the new arguments and spawn a new object based on the arguments that don't change. Example

```
base := &CmdContext {
  baseCmd:  "kubectl",
  baseURL:  "http://kubernetes.io/docs/user-guide/kubectl",
  In:       in,
  Out:      out,
  Errout:   errout,
}

base.New("get", "/kubectl_get")
base.New("create", "/kubectl_create")
base.New("delete", "/kubectl_delete/disclaimer")
...
```

**Comment by fabianofranz** [comment]
@kubernetes/kubectl

**Comment by AdoHe** [comment]
`CmdContext` sounds good to me. But I would not expect construct context every time we new a command, there should be some common context.

**Comment by dgonzalez** [comment]
Perfect. I am nearly done with the feedback and one example command. I will
push it soon and have another pass. TonyAdo there will be a New method for
constructing commands with the common attributes (which are the majority)
factored in.

2016-10-11 1:49 GMT+01:00 TonyAdo <notifications@github.com>:

**Comment by dgonzalez** [comment]
@fabianofranz I have updated the context. I'll work in a couple of test commands to see if this really works.

**Comment by dgonzalez** [comment]
I thought a bit about a funky structure for the sub commands but it really looks to me that a simple array will do the job without too much hassle.

File: pkg/kubectl/cmd/context.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+type CmdContext struct {
+	In            io.Reader
+	Out           io.Writer
+	ErrOut        io.Writer
+	BaseCmd       string
+	CmdName       string
+	FullName      string
+	CmdDetailsURL string
+	SubCmdNames   []string
```
Comment (deads2k) [inline pkg/kubectl/cmd/context.go]: I wasn't expecting to see this.  Can you try to plumb it through without this one?

The rest matches my expectations.

Comment (dgonzalez) [inline pkg/kubectl/cmd/context.go]: This is in order to fit the comment from @fabianofranz about commands with subcommands:
```
kubectl create quota
```
Or we actually can use the CmdName for setting the value `create quota` and forget about the `SubCmdNames`... I am very open on that

Comment (deads2k) [inline pkg/kubectl/cmd/context.go]: > Or we actually can use the CmdName for setting the value create quota and forget about the SubCmdNames... I am very open on that

I think the name becomes `ctx.BaseCmd + " " + ctx.CmdName`.  Using a `ctx.SubCommand("next-name")` will be able to set the new `BaseCmd` to existing `BaseCmd + " " + CmdName` and set the new `CmdName`.

**Comment by dgonzalez** [comment]
Partially solves #31267

File: pkg/kubectl/cmd/context.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+func (ctx CmdContext) New(CmdName string, CmdDetailsURL string) *CmdContext {
```
Comment (kargakis) [inline pkg/kubectl/cmd/context.go]: Do we need New as a method?

**Comment by dgonzalez** [comment]
We do because we don't want to create the command every time as it gets too
verbose. We could have also a new as a function but... it is the same in
the end isn't it?

2016-10-14 11:17 GMT+01:00 Michail Kargakis <notifications@github.com>:

**Comment by kargakis** [comment]
`New` as is is appropriate for chaining which means the name is misleading because I cannot create a new Context if I don't have one already.

File: pkg/kubectl/cmd/context.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	return ctx.NewForSub(CmdName, CmdDetailsURL, nil)
+}
+
+func (ctx CmdContext) NewForSub(CmdName string, CmdDetailsURL string, SubCmdNames []string) *CmdContext {
+	newCtx := &CmdContext{
```
Comment (deads2k) [inline pkg/kubectl/cmd/context.go]: I expected this work to be inlined in the `Sub` method.

File: pkg/kubectl/cmd/context.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	FullName      string
+	CmdDetailsURL string
+	SubCmdNames   []string
+}
+
+func (ctx CmdContext) New(CmdName string, CmdDetailsURL string) *CmdContext {
+	return ctx.NewForSub(CmdName, CmdDetailsURL, nil)
+}
+
+func (ctx CmdContext) NewForSub(CmdName string, CmdDetailsURL string, SubCmdNames []string) *CmdContext {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/context.go]: All variables and function parameters should be `camelCase`, not `CamelCase`

**Pull request closed without merging**